pkg/api/service.go

```diff
@@ -91,18 +164,23 @@ func (s *Service) SubscribeEnvelopes(
 	}
 
 	ch := s.subscribeWorker.listen(stream.Context(), query)
+	err = s.catchUpFromCursor(stream, query, log)
+	if err != nil {
+		return err
+	}
+
+	// TODO(rich): Pull from DB one more time after first tick of subscribe worker.
+
 	for {
 		select {
 		case envs, open := <-ch:
 			if open {
-				err := stream.Send(&message_api.SubscribeEnvelopesResponse{
-					Envelopes: envs,
-				})
+				err := s.sendEnvelopes(stream, query, envs)
 				if err != nil {
 					return status.Errorf(codes.Internal, "error sending envelope: %v", err)
 				}
 			} else {
-				// TODO(rich) Recover from backpressure
+				// TODO(rich) Reset whole subscribe flow from new cursor
 				log.Debug("channel closed by worker")
```

Comment:
If the channel gets filled up, I think handling it is fairly simple:

1. Drain the channel, updating the cursor as we do so.
2. Restart the whole subscription request starting from that cursor (setting up a new listener, and catching up from that cursor).

If the consumer on the stream is slow, then it just means that the stream will be permanently behind the database, and the consumer will be in the catch-up flow forever. The catch-up flow simply does a poll every 50 ms, blocking while it sends payloads down before the next poll.